torch/_dynamo/symbolic_convert.py

```diff
@@ -952,40 +946,35 @@ def COMPARE_OP(self, inst):
                     **options,
                 )
             )
-        elif (
-            isinstance(left, TensorVariable) or isinstance(right, TensorVariable)
-        ) and op in supported_tensors:
-            self.push(
-                wrap_fx_proxy(
-                    self,
-                    supported_tensors[op](left.as_proxy(), right.as_proxy()),
-                    **options,
-                )
-            )
-        elif (
-            isinstance(left, DynamicShapeVariable)
-            or isinstance(right, DynamicShapeVariable)
-        ) and op in supported_tensors:
-            self.push(
-                DynamicShapeVariable.create(
-                    self,
-                    supported_tensors[op](left.as_proxy(), right.as_proxy()),
-                    dyn_shape=None,
-                    **options,
-                )
-            )
         elif op in ("in", "not in"):
             self.push(right.call_method(self, "__contains__", [left], {}))
             if op == "not in":
                 self.UNARY_NOT(inst)
         elif (
-            isinstance(left, UserFunctionVariable)
-            and isinstance(right, UserFunctionVariable)
-            and op in supported_is_const
-        ):
-            self.push(
-                ConstantVariable(supported_is_const[op](left.fn, right.fn), **options)
+            # TODO: feels like we should be able to refactor so that VariableTracker.compare()
+            # covers more cases. However some of the cases are simpler to leave in the current function.
+            # (for example, if both left an right are python constants, always return a constant
+            # instead of re-writing that logic for every type of VariableTracker that can be constant)
+            isinstance(
+                left,
+                (
+                    TensorVariable,
+                    DynamicShapeVariable,
+                    BaseListVariable,
+                    UserFunctionVariable,
+                ),
             )
+            or isinstance(
+                right,
+                (
+                    TensorVariable,
+                    DynamicShapeVariable,
+                    BaseListVariable,
+                    UserFunctionVariable,
+                ),
+            )
+        ):
+            self.push(left.compare(self, op, right, **options))
```

Comment:
On second though, I'd suggest trying to reuse some existing machinery for this rather than adding a new `compare()` method.

We already have:

1)
```
left.call_method(tx, "__eq__", [right], {})
```

2)
```
BuiltinVariable(operator.eq).call_function(tx, [left, right], {})
```

Often option 2 will redispatch to option 1:
https://github.com/pytorch/pytorch/blob/0f802eedc2671fd17106923ef2981e3e6eaa8b95/torch/_dynamo/variables/builtin.py#L547-L563

This has the advantage that it will work if the user calls `__eq__` or `operator.eq` directly.

We should be able to just make ths:
```suggestion
        else:
            self.push(BuiltinVariable(supported_any[op], **options).call_function(tx, [left, right], {}))
```

Reply:
This doesn't change the suggestion, but @jansel I want to draw your attention to an ongoing conversation we're having at https://github.com/pytorch/pytorch/pull/93271 The short version:

1. For full fidelity operator.add handling, we need "reversibility" logic (that's what #93271 is about) to try both `__add__` and then try `__radd__` if add is NotImplemented
2. For FX graph tracing purposes, we prefer to trace into the graph at the operator.add level, not the `__add__` method call level, for a variety of reasons, but a very pedestrian reason is when you do reversibility logic it looks really ugly to add FX nodes for both `__add__` and `__radd__`

cc @jbschlosser I wonder if `BINARY_ADD` and similar opcodes go via the builtin variables

Reply:
Sure. We can do it through call_eq, call_gt, etc. However, you still need to have a per-object expansion of equality comparison - I am not sure there is a way of getting around .compare() calls or __eq__ calls or equivalent on the VariableTrackers themselves. 

So even if we rewrote this to be BuiltinVariable's roughly `call_{supported_any[op]}` - it would not preclude the need to compare in [left] [right] - in our case, which would still need to be delegated to ListVariable's `__eq__` or `compare` (same thing), which in turn would need to call that again on the contained (`.items`) objects. 

Would you agree with that?

Reply:
> This doesn't change the suggestion, but @jansel I want to draw your attention to an ongoing conversation we're having at #93271 The short version:
> 
> 1. For full fidelity operator.add handling, we need "reversibility" logic (that's what [General reversible binary op support (e.g. __add__ / __radd__) in dynamo #93271](https://github.com/pytorch/pytorch/pull/93271) is about) to try both `__add__` and then try `__radd__` if add is NotImplemented
> 2. For FX graph tracing purposes, we prefer to trace into the graph at the operator.add level, not the `__add__` method call level, for a variety of reasons, but a very pedestrian reason is when you do reversibility logic it looks really ugly to add FX nodes for both `__add__` and `__radd__`
> 
> cc @jbschlosser I wonder if `BINARY_ADD` and similar opcodes go via the builtin variables

I said this in chat - but if we do this, we might as well go all the way and break up the giant if/else trees in `call_method` and `call_function` with delegation to ops-defined-on-subclass-variable-trackers like this...

Reply:
@voznesenskym I think a helper method makes sense, but it should be local to list variable (not a new API on every single Variable Tracker)

@ezyang I'll make some comments over there (once I finish going through).  I wasn't trying to make a strong argument for function vs method in this case, more of a "don't and a third option"

Reply:
> @voznesenskym I think a helper method makes sense, but it should be local to list variable (not a new API on every single Variable Tracker)

Agreed, the `compare` barely do anything anyway, just grab the args and essentially make a new proxy for the comparison op.

Reply:
> cc @jbschlosser I wonder if BINARY_ADD and similar opcodes go via the builtin variables

They do:
https://github.com/pytorch/pytorch/blob/37f7c00a8aae2ced9666a69d0d51a4d9d4b379af/torch/_dynamo/symbolic_convert.py#L1462

https://github.com/pytorch/pytorch/blob/37f7c00a8aae2ced9666a69d0d51a4d9d4b379af/torch/_dynamo/symbolic_convert.py#L133-L141